cmd/relay/repomgr/repomgr.go

```diff
@@ -0,0 +1,398 @@
+package repomgr
+
+import (
+	"bytes"
+	"context"
+	"errors"
+	"fmt"
+	"log/slog"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	atproto "github.com/bluesky-social/indigo/api/atproto"
+	"github.com/bluesky-social/indigo/atproto/identity"
+	atrepo "github.com/bluesky-social/indigo/atproto/repo"
+	"github.com/bluesky-social/indigo/atproto/syntax"
+	"github.com/bluesky-social/indigo/cmd/relay/events"
+	"github.com/bluesky-social/indigo/cmd/relay/models"
+	"github.com/ipfs/go-cid"
+	"go.opentelemetry.io/otel"
+)
+
+const defaultMaxRevFuture = time.Hour
+
+func NewRepoManager(directory identity.Directory, inductionTraceLog *slog.Logger) *RepoManager {
+	maxRevFuture := defaultMaxRevFuture // TODO: configurable
+	ErrRevTooFarFuture := fmt.Errorf("new rev is > %s in the future", maxRevFuture)
+
+	return &RepoManager{
+		userLocks:         make(map[models.Uid]*userLock),
+		log:               slog.Default().With("system", "repomgr"),
+		inductionTraceLog: inductionTraceLog,
+		directory:         directory,
+
+		maxRevFuture:           maxRevFuture,
+		ErrRevTooFarFuture:     ErrRevTooFarFuture,
+		AllowSignatureNotFound: true, // TODO: configurable
+	}
+}
+
+func (rm *RepoManager) SetEventManager(events *events.EventManager) {
+	rm.events = events
+}
+
+// RepoManager is a poorly defined chunk of code
+// TODO: RepoManager should probably merge with what calls it or what it calls; probably move HandleCommit into bgs.go
+type RepoManager struct {
+	lklk      sync.Mutex
+	userLocks map[models.Uid]*userLock
+
+	events *events.EventManager
+
+	log               *slog.Logger
+	inductionTraceLog *slog.Logger
+
+	directory identity.Directory
+
+	maxRevFuture       time.Duration
+	ErrRevTooFarFuture error
+
+	// AllowSignatureNotFound enables counting messages without findable public key to pass through with a warning counter
```

Comment:
what is the use-case for this?

Reply:
PLC is down. New repo exists in PDS but hasn't forwarded DID to PLC yet?

Reply:
I think if PLC is down, and the identity isn't cached, the network is basically expected to pause? especially for firehose consumers, where they can catch up when it comes back up (presumably within minutes, PLC is critical infrastructure).